blockchain/indexbuilder.go

```diff
@@ -113,12 +119,145 @@ func (ib *IndexBuilder) HandleBlock(blk *block.Block) error {
+	tipIndexBytes := byteutil.Uint64ToBytes(tipIndex)
+	batch.Put(blockActionBlockMappingNS, indexActionsTipIndexKey, tipIndexBytes, "failed to put tip index of actions")
+	tipHeightBytes := byteutil.Uint64ToBytes(tipHeight)
+	batch.Put(blockActionBlockMappingNS, indexActionsTipHeightKey, tipHeightBytes, "failed to put tip height")
+	if err := ib.store.Commit(batch); err != nil {
+		return err
+	}
+	return nil
+}
+func (ib *IndexBuilder) initAndLoadActions() error {
+	err := initIndexActionsKey(ib.store)
+	if err != nil {
+		return err
+	}
+	tipHeight, err := ib.dao.getBlockchainHeight()
+	if err != nil {
+		return err
+	}
+	startHeight, startIndex, err := ib.getStartHeightAndIndex()
+	if err != nil {
+		return err
+	}
+	zap.L().Info("Loading actions", zap.Uint64("startHeight", startHeight), zap.Uint64("startIndex", startIndex))
+	batch := db.NewBatch()
+	i := startHeight
+	for ; i <= tipHeight; i++ {
+		hash, err := ib.dao.getBlockHash(i)
+		if err != nil {
+			return err
+		}
+		body, err := ib.dao.body(hash)
+		if err != nil {
+			return err
+		}
+		blk := &block.Block{
+			Body: *body,
+		}
+		err = indexBlockHash(startIndex, hash, ib.store, blk, batch)
+		if err != nil {
+			return err
+		}
+		receipts, err := ib.dao.getReceipts(i)
+		// if receipts are not available,this error will be ignored
+		if err != nil && errors.Cause(err) != db.ErrNotExist {
+			return err
+		}
+		putReceipts(i, receipts, batch)
+		startIndex += uint64(len(blk.Actions))
+		// commit once every 10000 heights
+		if i%10000 == 0 {
+			if err := ib.commitBatchAndClear(startIndex, i, batch); err != nil {
+				return err
+			}
+		}
+		if i%1000 == 0 {
+			zap.L().Info("committing actions", zap.Uint64("height", i), zap.Uint64("index", startIndex))
+		}
+	}
+	// last commit
+	zap.L().Info("last committing actions", zap.Uint64("height", i-1), zap.Uint64("index", startIndex))
+	if err := ib.commitBatchAndClear(startIndex, i-1, batch); err != nil {
+		return err
+	}
+	return nil
+}
+func getNextIndex(store db.KVStore) (uint64, error) {
+	value, err := store.Get(blockActionBlockMappingNS, indexActionsTipIndexKey)
+	if err != nil && errors.Cause(err) == db.ErrNotExist {
+		return 0, initIndexActionsKey(store)
+	}
+	NextIndex := enc.MachineEndian.Uint64(value)
+	return NextIndex, nil
+}
+func getNextHeight(store db.KVStore) (uint64, error) {
+	value, err := store.Get(blockActionBlockMappingNS, indexActionsTipHeightKey)
+	if err != nil && errors.Cause(err) == db.ErrNotExist {
+		return 0, initIndexActionsKey(store)
+	}
+	NextHeight := enc.MachineEndian.Uint64(value)
+	NextHeight++
+	return NextHeight, nil
+}
 func indexBlock(store db.KVStore, blk *block.Block, batch db.KVStoreBatch) error {
 	hash := blk.HashBlock()
-	for _, elp := range blk.Actions {
+	// get index that already builded
+	startIndex, err := getNextIndex(store)
+	if err != nil {
+		return err
+	}
+	// get height that already builded
+	startHeight, err := getNextHeight(store)
```

Comment:
It shouldn't come from the db. Instead, it should come from the current block's height.

Reply:
Or maybe you want to check if startHeight == blk.height

Reply:
startHeight alway not equal to blk.height,especially when diable plugin after enable plugin